### DISCUSSION TITLE
When does compilation into bytecode happen exactly?

### DISCUSSION THREAD

**Discussion by IslamRustamov** [facebook/hermes#1152] [Q&A] [answered]
So, let’s imagine that we are building Android app with React Native and Hermes enabled.
We know that Hermes compiles JS into bytecode, but when it happens exactly? When we build an APK, or when user installs this APK?

**Comment by ecreeth** [comment, marked as answer]
This happens at build time.  You can learn more about it [React Native EU 2019: Tzvetan Mikov - Improving React Native Performance
](https://www.youtube.com/watch?v=BkRE4MsqFXU)

**Reply by IslamRustamov** [reply]
Does build time include installation of the app on the phone?

**Reply by ecreeth** [reply]
No, as you can see in the image below, the app only loads and executes the bytecode.
![image](https://github.com/facebook/hermes/assets/20761166/ca1540c9-3411-4dd0-8097-7d1792cef69f)

**Reply by tmikov** [reply]
Yes, exactly.

**Reply by IslamRustam** [reply]
So, I tried to build an app with JSC and Hermes, reverse engineered APKs, looked at index.android.bundles and that's what I saw:
<img width="751" alt="Screenshot 2023-10-11 at 22 09 00" src="https://github.com/facebook/hermes/assets/71266873/0300a324-8a1a-4edb-8cb8-58bbd0dcbc2e">
<img width="725" alt="Screenshot 2023-10-11 at 22 09 07" src="https://github.com/facebook/hermes/assets/71266873/fc4d2909-a8f8-48d6-974e-4f6807bf2b2a">

The first one is with Hermes off, another one Hermes on. Do I understand correctly that this second example IS the bytecode (that cannot be viewed properly because of encoding)?

**Reply by ecreeth** [reply]
Yes. The bytecode cannot be viewed as plain text.

**Reply by IslamRustam** [reply]
Alright. Thanks!

Just the last 2 questions:
1. Does it mean that using Hermes is more secure than JSC? I mean, is there any way we can "debug" this bytecode/do something with it? (not that we could understand much from minified JS from JSC bundling, but still)
2. Probably a silly question, is there any way to "view" the bytecode? For example, we can view photo if it's in base64 format.

**Reply by ecreeth** [reply]
You can "read" a decompiled version of the bytecode with [hbctool](https://github.com/bongtrop/hbctool) or other tools like that. It doesn't mean it's more secure, it is just more difficult to read.

**Reply by IslamRustam** [reply]
Alright, thanks for that!

**Reply by tmikov** [reply]
also `hermes -dump-bytecode file.hbc`
